routing: Support Pathfinding to Blinded Routes

carlaKC opened the pull request:
## Change Description
This PR surfaces the ability to create routes to blinded path via the `QueryRoutes` api. It takes the approach of expressing a blinded path as a chain of hop hints, and backfills all the fields required for blinded payments in a second pass. 

Fixes #7200. 
Depends on [lightning-onion/57.](https://github.com/lightningnetwork/lightning-onion/pull/57)

## Steps to Test - LDK
Setup the following network + channels using my [branch of ldk-sample](https://github.com/carlaKC/ldk-sample/commit/94d1705e1f52bdc35cef42e605ca014b23280d5a): `LND --- LDK0 --- LDK1`

Create an invoice on LDK1: 
- `getinvoice (amt) (expiry seconds)` -> `lnr...`
- `lncli decodepayreq lnr...` -> get payment hash,  payment addr and cltv delta

Create a blinded path to LDK1 using tests: 
- Checkout [my branch of rust-lightning](https://github.com/carlaKC/rust-lightning/tree/route-blinding-interop) and open `lightning/src/blinded_path/test.rs`
- Update the following variables: 
  - `introduction_node` = LDK0 pubkey
  - `blinded_node` = LDK1 pubkey
  - `payment_secret` = payment addr from invoice
  - channel_id = channel connecting LDK0 --> LDK1
 - Run the test using `cargo test -- --nocapture test_blinded_path` -> this will print out a blinded path. 

Pay the blinded path via LND:
```
lncli queryroutes 
--introduction_point={cleartext_introduction_node_id} 
--blinding_point={blinding_point} 
--blinded_hops={blinded_introduction_node}:{introduction_node_data}
--blinded_hops={blinded_node}:{blinded_node_data}
--blinded_cltv= 120 + {invoice_cltv_delta}
--amt={invoice_amount} |
 lncli sendtoroute --payment_hash={payment_hash} -
```

The payment should succeed! 

Footguns: 
* CLTV:  make sure to add the delta in the LDK invoice to the value provided in the blinded path (which is just hardcoded to 120).
* Payment address: needs to be unique per-payment (for LDK reasons), so you have to regenerate every time.

## Steps to Test - CLN

<details>
  <summary>(this way is harder IMO, but was originally included in the PR description).</summary>
Since LND doesn't _produce_ blinded routes or forward blinded payments, you'll need the following setup: 
`Alice (LND) --- Bob(CLN) --- Carol (CLN) --privatechan-- Dave(CL)`

Run all CLN nodes in developer mode with: `--experimental-offers --dev-fast-gossip`.

You want an invoice with a blinded path, to do this you can create an offer and then fetch the raw invoice:
- `cl-dave offer 1000 test` -> `lno...` is your offer
- `cl-carol fetchinvoice lno...` -> `lni....` is your invoice
- `cl-carol decode lni...` -> This will give blinded hops/data etc. 

You can then pay the blinded invoice from LND: 
```
lncli queryroutes --introduction_point={cleartext_introduction_node_id} --blinding_point={blinding_point} --blinded_hops={blinded_introduction_node}:{introduction_node_data} --blinded_cltv={blinded_route_total_cltv} --amt={invoice_amount} | lncli sendtoroute --payment_hash={payment_hash} -
```

CLN only creates a blinded path when the node has private channels, and otherwise will just create an invoice with the receiving node as the introduction node. To test different variations of blinded paths: 
* With blinded hop: create offer on Dave and fetch with Carol
* With just introduction hop: create offer on Carol and fetch with Bob

</details>

carlaKC commented:
Opening this PR early to get feedback on the approach 🙏  As we open up more PRs for route blinding, this should be easier to test and form a more coherent piece of work.

bjarnemagnussen commented on routing/pathfind.go:
```diff
@@ -108,7 +109,8 @@ type finalHopParams struct {
 func newRoute(sourceVertex route.Vertex,
 	pathEdges []*channeldb.CachedEdgePolicy, currentHeight uint32,
-	finalHop finalHopParams) (*route.Route, error) {
+	finalHop finalHopParams, blindedPath *sphinx.BlindedPath) (
```
NIT: Maybe add a little bit of explanation about the `blindedPath` to the function documentation, i.e. that `blindedPath` is used to backfill the fields of the `pathEdges` for the part that is considered the blinded path. The commit message makes it very clear but from the function signature alone it may be confusing to understand if `pathEdges` already contain "blinded hops" or not?

bjarnemagnussen commented on lnrpc/lightning.proto:
```diff
@@ -2848,6 +2848,11 @@ message QueryRoutesRequest {
 
+    // An optional blinded path to reach the destination via. Note that the
+    // introduction node included in this path *must* be the same as the target
+    // pubkey provided, and there must be at least two hops in the path.
```
NIT: I think now with the introduction node allowed to be the destination of a blinded payment a single hop is allowed (see also description of `blinded_hops` under `BlindedRoute` below).

bjarnemagnussen commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -258,7 +256,78 @@ func (r *RouterBackend) parseQueryRoutesRequest(in *lnrpc.QueryRoutesRequest) (
+		blindedPmt     *routing.BlindedPayment
+	)
+
+	// If a blinded path is included, validate that values in the
```
NIT: Sentence of comment is not completed.

bjarnemagnussen reviewed (commented):
Just leaving some smaller comments for things I fell over while taking a glimpse at it.

carlaKC commented on routing/blinding.go:
```diff
@@ -0,0 +1,79 @@
+	// ErrInsufficientBlindedHops is returned when a blinded path does
+	// not have enough blinded hops.
+	ErrInsufficientBlindedHops = errors.New("blinded path requires " +
+		"at least two hops")
```
Nice catch, will update the error's wording 👍

carlaKC commented on routing/router.go:
```diff
@@ -1741,16 +1741,86 @@ type routingMsg struct {
+	Source route.Vertex
+
+	// Target is the node that the path terminates at.
+	Target *route.Vertex
```
This is in preparation for a future commit where we want to allow a nil `Target` value in the case where the end user has provided a blinded path (where we can get the target from the final hop). 

This is a somewhat stylistic choice, because we could also ask the user to add a target *and* a blinded route (and then check that they're equal) _but_ that seemed like more code + more user hassle so I went with mutually exclusive instead.

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@bitromortac: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@bitromortac: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@bitromortac: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@bitromortac: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@carlakc, remember to re-request review from reviewers when ready

carlaKC commented on routing/route/route.go:
```diff
@@ -147,6 +156,11 @@ func (h *Hop) Copy() *Hop {
 		c.AMP = &a
 	}
 
+	if h.BlindingPoint != nil {
```
`c := *h` will deep copy byte slices - https://go.dev/play/p/1dXJ3VZuvFF

bitromortac replied:
Hm, I'm not sure, if you change line 36 to `b.encryptedData[2] = 1`, it modifies `a` as well? Here is another minimal example https://go.dev/play/p/DmOngBuziEG

carlaKC replied:
Sure, but that example doesn’t include a dereference that does the copy for us?

https://stackoverflow.com/questions/38443348/does-dereferencing-a-struct-return-a-new-copy-of-struct

Byte slices are considered value types (per https://go.dev/blog/slices, passing slices to functions) so they’ll be copied in ‘c = *h’

carlaKC commented on lntest/itest/lnd_route_blinding.go:
```diff
@@ -0,0 +1,291 @@
+		bob   = ht.Bob
+	)
+
+	// Setup a two hop channel network: Alice -- Bob -- Carol.
```
I don't see how it would _increase_ coverage of pathfinding to a blinded route to cover _beyond_ the introduction node?

Also, how would you add a receiver node when LND doesn't support receiving?

bitromortac replied:
Yes, good point, right now it doesn't make sense :+1:

carlaKC commented on lntest/itest/lnd_route_blinding.go:
```diff
@@ -0,0 +1,291 @@
+	} else {
+		policy = bobCarolInfo.Node2Policy
+	}
+	require.Equal(ht, bobCarolBase, uint64(policy.FeeBaseMsat), "base fee")
```
Cool!, I clearly have to catch up with all the cool new `require` functionality!

carlaKC commented on lnrpc/lightning.proto:
```diff
@@ -2848,6 +2848,10 @@ message QueryRoutesRequest {
 
+    // An optional blinded path to reach the destination. Note that the
+    // introduction node must be provided as the first hop in the route.
+    BlindedPayment blinded_path = 19;
```
It is indeed possible to include multiple blinded paths, though this PR wouldn't be able to support pathfinding to multiple (see "Conveniently Out of Scope" in tracking issue 😅 ). 

For now, I'm going to update this to allow multiple blinded paths but restrict length to 1. 
Once we're all masters of our blinded path craft, we can lift that restriction to allow multiple paths.

Nice catch! Would have been a sticky API choice to work around without breaking in the future.

carlaKC commented:
Thanks for taking a look! Rebased on master, addressed comments and updated to the latest and greatest lightning-onion version. 

> think that there may be some overlap with https://github.com/lightningnetwork/lnd/pull/6920 (this is @joostjager's idea to unify route construction by letting pathfinding do all the work and to remove newRoute

I'm reticent to make this a dependency of the refactor (since the forwarding PR already depends on this), but I'll keep track of it + rebase as required!

carlaKC commented on record/hop.go:
```diff
@@ -17,6 +18,14 @@ const (
+
 	// MetadataOnionType is the type used in the onion for the payment
 	// metadata.
 	MetadataOnionType tlv.Type = 16
```
Afaik this field is only for recipients, so we won’t need to include it until LND supports receiving (I’d say correct failure is to error unknown if we get a payload with it).

carlaKC commented on routing/blinding.go:
```diff
@@ -0,0 +1,80 @@
+
+// AggregateRelay represents the aggregate payment relay parameters for a
+// blinded payment.
+type AggregateRelay record.PaymentRelayInfo
```
See commit message / follow up PR - I think the natural place for this to live is record but we have the same fields in different contexts.

carlaKC commented on record/blinded_data.go:
```diff
@@ -0,0 +1,25 @@
+}
+
+// PaymentConstraints is a set of restrictions on a payment.
+type PaymentConstraints struct {
```
I think that'll make the names _a little_ too long?

ziggie1984 replied:
Q: I still do not understand why we have the HtlcMin here and why not HtlcMax too, thats a spec question but maybe you know it ?

ziggie1984 commented on routing/route/route_test.go:
```diff
@@ -188,6 +188,37 @@ func TestPayloadSize(t *testing.T) {
 			},
 			Metadata: []byte{10, 11},
 		},
+		{
```
I think if we wanna mirror a payload-size with the "normal" blinded route, we should define the Introduction point (meaning that it should have the blinded-point in its onion) and the last hop should not have the blinded point in it, because it receives in the update_htlc record and the above hop would need a channelid too because it's not the last hop anymore?

ziggie1984 commented on routing/blinding.go:
```diff
@@ -78,3 +80,87 @@ func (b *BlindedPayment) Validate() error {
+// path.
+func (b *BlindedPayment) toRouteHints() map[route.Vertex][]*channeldb.CachedEdgePolicy { //nolint:lll
+	// If we just have a single hop in our blinded route, it just contains
+	// an introduction node (this is valid per the specification). Since
```
Nit: (this is valid per ~~the~~ specification)

ziggie1984 commented on routing/route/route_test.go:
```diff
@@ -160,6 +160,20 @@ func TestAMPHop(t *testing.T) {
 	}
 }
 
+// TestNoForwardingParams tests packing of a hop payload without and amount or
```
nit: `payload without an amount or expiry height`

ziggie1984 replied:
still and instead of `an` ?

ziggie1984 commented on routing/pathfind_test.go:
```diff
@@ -3130,3 +3131,186 @@ func dbFindPath(graph *channeldb.ChannelGraph,
+// following topology:
+//
+// A -- B -- C (introduction point) - D (blinded) - E (blinded).
+func TestBlindedRouteConstruction(t *testing.T) {
```
Hmm I find this testcase misses the whole fee/cltv calculation for the blinded paths, is this tested elsewhere, if so I would prefer a comment describing what's actually tested here, meaning that we are just nulling the values of the route, all the fee and cltv stuff for blinded routes is not tested here.

ziggie1984 commented on routing/pathfind_test.go:
```diff
@@ -3130,3 +3131,186 @@ func dbFindPath(graph *channeldb.ChannelGraph,
+		_, alicePk = btcec.PrivKeyFromBytes([]byte{1})
+		_, bobPk   = btcec.PrivKeyFromBytes([]byte{2})
+		_, carolPk = btcec.PrivKeyFromBytes([]byte{3})
+		_, davePk  = btcec.PrivKeyFromBytes([]byte{4})
```
Nit: shouldn't this be daveBlindedPk

ziggie1984 commented on routing/pathfind_test.go:
```diff
@@ -3130,3 +3131,186 @@ func dbFindPath(graph *channeldb.ChannelGraph,
+		// Create edges for the blinded portion of the route with no
+		// fee values, as would be the case when we construct a route
+		// from blinded-route-generated hop hints.
+		carolDaveEdge = &channeldb.CachedEdgePolicy{
```
Shouldn't this edge include the aggregate conditions for the blinded route, mimicking the behaviour of findPath where we construct the edgepolicies via toRouteHints()

ziggie1984 commented on routing/pathfind_test.go:
```diff
@@ -3130,3 +3131,186 @@ func dbFindPath(graph *channeldb.ChannelGraph,
+				eveHop,
+			},
+			// Current height = 100
+			// + 50 (blinded portion)
```
Nit: Thats not really the blinded portion, thats the final hop which is different from the blinded portion which is 0 in your testcase

ziggie1984 commented on lnrpc/lightning.proto:
```diff
@@ -3002,6 +3002,16 @@ message Hop {
     // The payment metadata to send along with the payment to the payee.
     bytes metadata = 13;
+
+    // Blinding point is an optional blinding point included for introduction
```
nit: maybe add that it's mandatory if this is an introduction node this hop?

ziggie1984 commented on lnrpc/lightning.proto:
```diff
@@ -3002,6 +3002,16 @@ message Hop {
+    // nodes in blinded paths.
+    bytes blinding_point = 14;
+
+    // Encrypted data is a receiver-produced blob of data that provides hops
```
Q:does this also account for the case where the sender and receiver combine blinded paths to hide the introduction point for example, then part of the encrypted data would be generated by the sender?

ziggie1984 commented on lnrpc/lightning.proto:
```diff
@@ -3357,6 +3361,62 @@ message RouteHint {
+    // The minimum hltc size for the route.
+    uint64 min_htlc = 1;
+
+    // The cltv limit for the route.
```
Nit: Isn't this just the limit for the blinded part of the route?

ziggie1984 commented on lnrpc/lightning.proto:
```diff
@@ -3357,6 +3361,62 @@ message RouteHint {
+    // The unblinded pubkey of the introduction node for the route.
+    bytes introduction_node = 1;
+
+    // The ephemeral pubkey used by nodes in the blinded route.
```
Q: what if we combine a receiver and sender blinded route, do these fields still need to be set?

ziggie1984 commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -258,7 +256,85 @@ func (r *RouterBackend) parseQueryRoutesRequest(in *lnrpc.QueryRoutesRequest) (
+				"different to blinded path: %v", in.CltvLimit,
+				blindedLimit)
+		}
+		cltvLimit = blindedLimit
```
so my understanding says that this blindedPath.RelayConstraints.CltvLimit should refer only to the limit inside the blinded path? Though why should we limit our CLTV-Limit for the other part of the route (the non-blinded part)?

carlaKC replied:
I think this was based on an old version of the spec - no cltv limit for blinded path payments in the merged version, now they're _just_ in the payloads for the blinded hops so that they can "expire" the blinded route to prevent probing.

ziggie1984 replied:
ohh thanks for the explanation now it makes sense

ziggie1984 replied:
Ok I have 2 question then:

1. Where do we get a reasonable `TotalCltvDelta` for the BlindedPayment, in prior Spec Version it seemed to me it was part of the Aggregated Data but now it seems we decide it by ourself? Feels not quite right or ?

carlaKC replied:
1. Total cltv delta for the blinded hops and blinded destination = `QueryRoutesRequest.BlindedPath.TotalCltvDetla`* + `QueryRoutes.Request.FinalCltvDelta` (The `FinalCltvDelta` would be obtained from the invoice containing the blinded path. 
2. This is already handled in pathfinding - the [finalHTLCExpiry](https://github.com/carlaKC/lnd/blob/bdb41e58677836390f817b481fd784148f310571/routing/router.go#L1778) is [provided to pathfinding](https://github.com/carlaKC/lnd/blob/bdb41e58677836390f817b481fd784148f310571/routing/router.go#L1793) and [added](https://github.com/lightningnetwork/lnd/blob/e549096b8837a95e0515c8c63fa01c634714d913/routing/pathfind.go#L613) to our `cltvLimit`.

*Looking at this again - I think that `TotalCltvDelta` in `BlindedPayment` is a bad name - `BlindedCltvDelta`?

ziggie1984 commented on itest/lnd_route_blinding.go:
```diff
@@ -0,0 +1,298 @@
+// that uses Carol as the introduction node (plus dummy hops to cover multiple
+// hops). The test simply asserts that the structure of the route is as
+// expected.
+func testQueryBlindedRoutes(ht *lntest.HarnessTest) {
```
Nice Itest 🤓🎉

ziggie1984 commented on itest/lnd_route_blinding.go:
```diff
@@ -0,0 +1,298 @@
+	var (
+		aggregateBaseFee   uint64 = 1500
+		aggregateCltvDelta uint32 = 125
+		cltvLimit                 = info.BlockHeight + 500
```
I think cltvLimit is a delta so no need to add blockheight, at least that's what `maxTotalTimelock` (which we take when no cltvLimit is set) is, which is default 2016 blocks

ziggie1984 reviewed (commented):
Awesome work Carla, feels close to have Route-Blinding in LND :) 🚀

carlaKC commented:
Cleaned up tests a bit + removed alias (was more confusing than saving the LOC was worth, and the spec has changed since writing - _slightly_ different fields now).

I want to run through and add more test coverage in a few places (suggestions for where coverage would be useful would be great!), but I don't think the structure will change much beyond that.

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@carlakc, remember to re-request review from reviewers when ready

ellemouton commented on routing/blinding.go:
```diff
@@ -67,3 +69,89 @@ func (b *BlindedPayment) Validate() error {
+
+	// Start at an offset of 1 because the first node in our blinded hops
+	// is the introduction node and terminate at the second-last node
+	/// because we're dealing with hops as pairs.
```
nitty nit: extra `/`

ellemouton commented on routing/blinding.go:
```diff
@@ -67,3 +69,89 @@ func (b *BlindedPayment) Validate() error {
+			ToNodeFeatures: features,
+		}
+
+		hints[fromNode] = []*channeldb.CachedEdgePolicy{
+			hint,
+		}
```
still making my way through, but currently am unsure about how we maintain the order of the hops in the blinded path if we are putting the hints in a map here 🤔

ellemouton replied:
ah ok i think i got it: for each entry, the `ToNodePubKey` will than have an entry in the `hints` which will then link the the next one. ie, the "chain".

carlaKC replied:
Yeah, product of how we express hint chains in a map. So `A -> B -> C` has map entries:
`A: ToNodePubkey() = B`
`B: ToNodePubkey() = C`

ellemouton commented on routing/router.go:
```diff
@@ -1793,24 +1799,78 @@ func NewRouteRequest(source route.Vertex, target *route.Vertex,
+			)
+		}
+
+		return route.NewVertex(hops[len(hops)-1].NodePub)
```
ok so is this where we would insert & return a dummy final target for multiple blinded routes?

carlaKC replied:
Indeed! I'd think that the actually switcheroo (putting a marker pubkey as the last hop in the blinded path) would happen in `toRouteHints()` but this would be where we return that marker pubkey so that pathfinding knows where to go.

ellemouton commented on routing/route/route.go:
```diff
@@ -385,6 +385,42 @@ func (r *Route) Copy() *Route {
 
 // HopFee returns the fee charged by the route hop indicated by hopIndex.
+//
+// This calculation takes into account the possibility that the route contains
```
excellent comment 💥

ellemouton commented on lnrpc/lightning.proto:
```diff
@@ -2891,6 +2891,10 @@ message QueryRoutesRequest {
     */
     repeated lnrpc.RouteHint route_hints = 16;
 
+    // An optional blinded path(s) to reach the destination. Note that the
```
nicro nit to use the `/*` comment pattern used in this file

ellemouton commented on lnrpc/lightning.proto:
```diff
@@ -2891,6 +2891,10 @@ message QueryRoutesRequest {
 
+    // An optional blinded path(s) to reach the destination. Note that the
+    // introduction node must be provided as the first hop in the route.
+    repeated BlindedPayment blinded_path = 19;
```
wait so we _are_ adding support for path finding to multiple blinded routes here? or am i missing something?

ellemouton replied:
the naming is also a bit confusing. Since this is called `BlindedPath` but is an array of `BlindedPayments` each of which include a `BlindedPath`

carlaKC replied:
> the naming is also a bit confusing. Since this is called BlindedPath but is an array of BlindedPayments each of which include a BlindedPath

naming 🙇‍♀️

Reasoning here was: 
- `BlindedPayment`: this _isn't_ a blinded path, since blinded paths can exist without payment params 
- `blinded_path`: calling it `blinded_payments` feels like we're making multiple payments which is also wrong?

Going to go for `repeated BlindedPaymentPath blinded_payment_paths` and pray for my line wrapping?

ellemouton commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -258,11 +256,75 @@ func (r *RouterBackend) parseQueryRoutesRequest(in *lnrpc.QueryRoutesRequest) (
+				in.PubKey)
+		}
+
+		if len(in.BlindedPath) != 1 {
+			return nil, errors.New("query routes only supports " +
+				"a single blinded path")
+		}
```
ah ha! gotcha 🤓

ellemouton reviewed (commented):
Beautiful code ⚡  Did my first, high level pass just to get an initial understanding. So far, everything looks good!

carlaKC commented on routing/blinding.go:
```diff
@@ -0,0 +1,69 @@
+
+// BlindedPayment provides the path and payment parameters required to send a
+// payment along a blinded path.
+type BlindedPayment struct {
```
We already have a [CLTVLimit](https://github.com/lightningnetwork/lnd/blob/f1bca2a59c6117cd2a23895d6b9c9a339904468b/routing/pathfind.go#L336) in our routing restrictions, so I was thinking that we just re-use that? 

There is a bit of a design choice in here about how we want to deal with the split between blinded payments and regular ones where there is field overlap: 

1. Only add additional blinded payment info, then re-use common fields (as is done here)
2. Have a completely separate set of fields, ensure they are mutually exclusive then merge

I've found (1) to be a _little_ more simple when it gets down the the API design for these (two fields in different places are confusing), so that's what swayed me in the end.

carlaKC commented on routing/blinding.go:
```diff
@@ -0,0 +1,69 @@
+
+// BlindedPayment provides the path and payment parameters required to send a
+// payment along a blinded path.
+type BlindedPayment struct {
```
Also, a spec note: there's still some ongoing discussion around the [specifics relating to bolt12 invoices](https://github.com/lightning/bolts/pull/1069). Doesn't matter to us since we're doing it in reverse, but wanted to flag!

carlaKC commented on routing/blinding.go:
```diff
@@ -0,0 +1,69 @@
+		return ErrNoBlindedPath
+	}
+
+	// The sphinx library inserts the introduction node as the first hop,
+	// so we expect at least one hop.
+	if len(b.BlindedPath.BlindedHops) < 1 {
```
As discussed offline, seems like the normal way to do it!

carlaKC commented on routing/blinding_test.go:
```diff
@@ -53,3 +57,106 @@ func TestBlindedPathValidation(t *testing.T) {
+	t.Parallel()
+
+	var (
+		_, pk1  = btcec.PrivKeyFromBytes([]byte{1})
+		_, pkb1 = btcec.PrivKeyFromBytes([]byte{2})
```
Would they? `pk1` is the introduction node and `pkb1` is the [blinded introduction node](https://github.com/lightningnetwork/lightning-onion/pull/57/files#diff-6883260b5ee8afaff7c13850ceede938210b4390bdf049068419912202c214beR89)  - the value will be different even though we don't use it. 

I specifically wanted different values so that a test would fall over if we're using the wrong thing.

carlaKC commented:
There are various cltv values flying around, so I thought that it would be worthwhile to give an overview so that we're all on the same page. 

Aggregate Parameters/CLTV Delta: 
* When a sender is provided with a blinded path, they also get the aggregate parameters for payments to the blinded path, including the   total `cltv_delta` for the blinded portion of that path.
* This `cltv_delta` represents the timelock from the introduction node  to the final node, and is *inclusive* of the receiving node's final   CLTV value.

Encrypted Data/CLTV Delta (not relevant for this PR): 
* The `outgoing_cltv_value` for hops *within* a blinded path is set  to zero by the sender, because they do not know what delta to apply for each node (only the total delta from the blinded path).
* Forwarding nodes will decrypt their `encrypted_data`, obtain their  `cltv_delta`, check that it aligns with their policy and then forward  the htlc with:  `outgoing_cltv_value` = incoming HTLC exipry - `ctlv_delta`.

CLTV Limit: 
* The total timelock (expressed as a delta from the current height) that LND allows a route to have. 
* By default, this value is 2016 blocks (DefaultMaxOutgoingCltvExpiry),but it can be specified via RPC. 

Blinded Routes / Max CLTV Expiry:
* Blinded routes have a `max_cltv_expiry` field that is set in its `encrypted_data` for each hop to stop probing. This value is chosen by the receiver, and known to each forwarding hop.
* LND already has logic in place that allows us to limit a route to a maximum total timelock, so we don't need to implement something new for blinded paths. 
* This value is *not* included in the blinded path struct because it will be communicated elsewhere (ie, in an invoice that contains a  blinded path).

Final CLTV Delta: 
* For bolt 11 invoices, we get a `min_final_cltv_expiry_delta` in the invoice, which the receiving node enforces to ensure that they have sufficient time to resolve the HTLC on chain.
* By contrast, blinded paths allow the receiver to "bake in" their final cltv expiry delta, because the sender must adhere to the aggregate delta (which includes their delta).

Final outgoing_expriy_delta: 
* While the blinded path will ensure that the receiving node gets the delta that it wants in the final HTLC, the sending node still needs to know what value to set for the outgoing_expriy in the receiving node's onion (as the spec indicates that the sender MUST include this value for the final hop in a blinded route).
* The spec recommends adding a random offset to the final node's final CLTV so that decreasing CLTVs for the payment can't trivially be observed and compared with the network topology to infer the location within the route. 
* As outlined in the current spec, we should set the outgoing_expiry to current_height + this sender_delta. 
* LND currently adds 3 blocks BlockPadding to the final CLTV delta for payments, but not for query routes / send to route.

For blinded paths: 
- Do not allow a final CLTV delta to be specified for blinded paths (ie, the value must be zero and we pass a zero value in). This will result in LND setting the last hop's outgoing_expiry to the current height for query routes. 
- For payments to blinded routes, this will need to be different: we will add the block padding and set current height + padding as the final outgoing_expiry.

carlaKC commented:
Addressed all the feedback so far here! 

Just going to do some inter-opping to make sure everything is in order then I'll re-request, but if folks want to take an early peek it's ready for a look.

ziggie1984 commented:
> This cltv_delta represents the timelock from the introduction node to the final node, and is inclusive of the receiving node's final CLTV value.

I don't think it is inclusive because we as a sender need to set the correct `outgoing_cltv_value ` for the final hop which means we as a sender do decide this value or. And in this PR this is always set to the `current_blockheight` which IMO should be padded to make it more secure?

ziggie1984 commented:
I have one major question here why do we not want to let the receiver specify a final_cltv_delta as in bolt11 and why does it not apply for blinded routes. I don't think the CLTV_DELTA is baked into the aggregate values of the blinded path.

ziggie1984 reviewed (commented):
Had some final comments, still struggling with the concept of the Final_CLTV_Delta for the final hop when blinded paths are active

carlaKC commented on routing/route/route.go:
```diff
@@ -396,8 +432,25 @@ func (r *Route) HopFee(hopIndex int) lnwire.MilliSatoshi {
-	return incomingAmt - r.Hops[hopIndex].AmtToForward
+	outgoingAmt := r.Hops[hopIndex].AmtToForward
+
+	switch {
```
Pre-existing assumption that we're calculating fees for a valid hop, not sure it's worth adding an error return here.

ziggie1984 replied:
agree I am ok without it too.

carlaKC commented on routing/pathfind_test.go:
```diff
@@ -3130,3 +3131,222 @@ func dbFindPath(graph *channeldb.ChannelGraph,
+		// Create channel edges for the unblinded portion of our
+		// route. Proportional fees are omitted for easy test
+		// calculations, but non-zero base fees ensure our fee is
+		// still accounted for. Feature vectors are left empty because
```
Empty as in nothing else in them, updating comment.

carlaKC commented on lnrpc/lightning.proto:
```diff
@@ -3380,6 +3392,53 @@ message RouteHint {
+    // The proportional fee for the blinded path provided, expressed in msat.
+    uint64 proportional_fee_msat = 3;
+
+    // The total CLTV delta for the blinded path provided, *not* including the
```
Outdated - nice catch! Updating comment accordingly.

carlaKC commented on docs/release-notes/release-notes-0.17.0.md:
```diff
@@ -93,6 +93,9 @@ package](https://github.com/lightningnetwork/lnd/pull/7356)
 * [Fixed a bug where we didn't check for correct networks when submitting
   onchain transactions](https://github.com/lightningnetwork/lnd/pull/6448).
 
+* [Query routes](https://github.com/lightningnetwork/lnd/pull/7267) has been 
```
I'll come back and write more in depth release notes once review is finalized, easily to do it at the end rather than constantly update them 👍

yyforyongyu replied:
cool think we can update this now

ziggie1984 commented:
My thoughts on the CLTV topic, especially when having a blinded path.

Currently when Lnd creates an Invoice we have a minimum when we specify it manually for Bolt11 invoices of [18](https://github.com/lightningnetwork/lnd/blob/aa17c17f9a019c44b44b5f57e3b00836deff7bbd/lnrpc/invoicesrpc/addinvoice.go#L366). Thats when we create the invoice.

Then we have another check in place, where we do fail the payment when the `outgoing_cltv_value` we receive does fall below the range of the [`currentheight + rejectdelta`](https://github.com/lightningnetwork/lnd/blob/aa17c17f9a019c44b44b5f57e3b00836deff7bbd/invoices/update.go#L223) which is currently 13 blocks.

That's on the lnd receive and create invoice side. As Carla mentioned correctly we also have a check in place when sending a payment, we will add a padding of 3 to the provided finalCLTVDelta in the invoice when sending a payment to an invoice.
(bolt11 currently).

Also Carla mentioned that that when sending to a route lnd will add nothing because it anticipates that the provided route is bulletproof.

Now when talking about Blinded Paths I would suggest we keep this FinalCLTV Requirement for the target node or we will have some exceptions to make when receiving to blinded paths for lnd. In addition we would have no way as a creator of the Invoice to hide us from the second-last node, we will depend on the sender choosing a reasonably high `sender_final_cltvdelta`(currently 3 blocks padding in this PR I think). In addition if we decide to use blinded paths also with Bolt11 invoice, we would not use the cltvDelta according to the current code (just theoretically speaking)?

In this PR we set the target `outgoing_cltv_value` to the currentblockheight + padding, which the second-last node would use to bound its outgoing htlc to knowing exactly that the next node is the receiver of the payment.

Apart from this topic this PR is good to go from my point of view

ziggie1984 commented on lnrpc/lightning.proto:
```diff
@@ -2911,12 +2914,21 @@ message QueryRoutesRequest {
     */
```
nit: maybe add a comment that they are not allowed when a blinded path is selected?

ziggie1984 commented on routing/pathfind.go:
```diff
@@ -819,6 +819,7 @@ func findPath(g *graphParams, r *RestrictParams, cfg *PathFindingConfig,
 				lnwire.TLVOnionPayloadOptional,
 			)
 
+			// TODO: this is now incorrect?
```
Interesting Point, I think we need to add the encrypted data here to make sure it does not exceed the maximum sphinx size?

carlaKC replied:
I'm tempted to fix this in a follow up (because touching findpath is scary), and we're already _technically_ incorrect here because we don't include things like metadata / MPP fields (and in practice we're nowhere near hitting this limit). WDYT?

ziggie1984 replied:
Sounds good, let's fix it in the follow-up PR.

ziggie1984 reviewed (commented):
Came across some final nits/questions, apart for that this things looks really good 👍

carlaKC commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -258,34 +256,136 @@ func (r *RouterBackend) parseQueryRoutesRequest(in *lnrpc.QueryRoutesRequest) (
+		// finalCLTVDelta varies depending on whether we're sending to
+		// a blinded route or an unblinded node. For blinded paths,
+		// our final cltv is already baked into the path so we restrict
+		// this value to zero on the API. Bolt11 invoices have a 
```
There is no specification for including blinded paths in bolt 11 invoices. If such a thing is introduced, then the code would need to be updated to reflect that.

ziggie1984 replied:
sgtm

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@ellemouton: review reminder
@bitromortac: review reminder
@carlakc, remember to re-request review from reviewers when ready

carlaKC commented on routing/pathfind_test.go:
```diff
@@ -3130,3 +3131,222 @@ func dbFindPath(graph *channeldb.ChannelGraph,
+		}
+
+		// Create final hop parameters for payment amount = 110. Note
+		// that final cltv delta is not set because blinded paths
```
It is allowed to be set (we can set a `sender_delta`), will update the comment to make a note of that.

carlaKC commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -258,34 +256,136 @@ func (r *RouterBackend) parseQueryRoutesRequest(in *lnrpc.QueryRoutesRequest) (
+				"zero for blinded paths")
+		}
+
+		// Just apply block padding for blinded routes, as CLTV delta 
```
This has to do with the block height that these expiries are expressed relative to - it's not actually a blinded paths issue, can happen with regular payments as well. Say we have: 
- Current Height = 100
- Final CLTV Expiry = 10 

### A) Same View of Chain
**Sender**: 
`cltv_expiry` for final incoming HTLC = `sender_current_height` + `final_cltv` + `sender_delta`
`cltv_expiry` for final incoming HTLC = 100 + 10 + 0 = 110

**Receiver**: 
Validate: `cltv_expiry >= receiver_current_height + min_final_cltv_expiry_delta` 
 Validate ✅: 110 >= 100 + 10  

### B) Different View of Chain (no delta)
Now, consider the case where the sender creates the route at height 100, sends the payment and by the time the recipient gets it we're at block 101.

**Sender**: 
`cltv_expiry` for final incoming HTLC = `sender_current_height` + `final_cltv` + `sender_delta`
`cltv_expiry` for final incoming HTLC = 100 + 10 + 0 = 110

**Receiver**: 
Validate: `cltv_expiry >= receiver_current_height + min_final_cltv_expiry_delta`
Validate ❌:  110 >= 101 + 10
-> validation will fail because it's comparing to a different current height!

For this reason, we want to set a `sender_delta` to allow for a block being mined (or the possibility that our view of the chain is a little behind) so that our payments still go through.  

### C) Different View of Chain (with delta)
For the same example above (B) where we send and receive with different block heights:

**Sender**: 
`cltv_expiry` for final incoming HTLC = `sender_current_height` + `final_cltv` + `sender_delta`
`cltv_expiry` for final incoming HLTC = 100 + 10 + 3 = 113

**Receiver**: 
Validate: `cltv_expiry >= receiver_current_height + min_final_cltv_expiry_delta` 
Validate ✅: 113 >= 101 + 10 

You can think of `final_cltv_delta` and the recipient saying "this is the number of blocks that I want from *my block height when I receive the HTLC* to be able to settle on chain", so we always need to make sure we give them _at least_ that delta.

carlaKC replied:
What's also pretty confusing here is that we don't add this `sender_delta` (/block padding) in the regular case of query routes (I'm not sure why - IMO it should have it, could be done separately).

ziggie1984 replied:
excellent explanation  🙏, I think it finally clicked. 

The main point is that the `min_final_cltv_expiry_delta ` and the `final_cltv` are different values for lnd.  

The `min_final_cltv_expiry_delta` for a lnd node is 13:
https://github.com/lightningnetwork/lnd/blob/e8c11b9898890857275fc25bd575fa0cace5db35/lncfg/config.go#L41

whereas the minimum `final_cltv_delta` we enforce is 18:

https://github.com/lightningnetwork/lnd/blob/e8c11b9898890857275fc25bd575fa0cace5db35/lnrpc/invoicesrpc/addinvoice.go#L366-L370

so the padding is baked in I guess. Maybe we also stick to this design for blinded paths as well. Meaning that we make sure when creating the blinded path that we do add a `blinded_path_cltv_delta` which enforces a `final_cltv_delta` of 18 and keep the `min_final_cltv_expiry_delta` at 13. 

One thing to keep in mind is that what happens to node_implementations when the `outgoing_cltv_value ` < `current_blockheight_sender`, normally for blinded paths this value should just adhere to the following requirement:

`cltv_expiry - cltv_expiry_delta >= outgoing_cltv_value` this will always be true if we include the `final_cltv_delta` in the `blinded_path_cltv_delta`.

But maybe some other implementations do check `outgoing_cltv_value` > `current_block_height` which is not part of the spec. So we could just add a 3 block padding to be absolutely sure. I am ok with both ways, as currently is or removing it and just let the creator of the blinded path assure it.

carlaKC replied:
> The main point is that the min_final_cltv_expiry_delta  and the final_cltv are different values for lnd.

Not sure I understand this statement? Both of these values are used for _receiving_, and this PR is only about sending..

> The min_final_cltv_expiry_delta for a lnd node is 13
> whereas the minimum final_cltv_delta we enforce is 18:

There's a difference between: 
- `min_final_cltv_expiry_delta`: the lowest CLTV delta we'll accept for an incoming HLTC (mostly spontaneous payments)
- `final_cltv_delta`: the CLTV delta we set for our invoices.

I think that the naming is tripping you up here. In the spec we [refer](https://github.com/lightning/bolts/blob/master/04-onion-routing.md#payload-for-the-last-node) to `min_final_cltv_delta` which is _actually_ `final_cltv` (ie, the value in the [invoice](https://github.com/lightning/bolts/blob/master/11-payment-encoding.md#tagged-fields)).

> so the padding is baked in I guess

The padding is *not* baked in. The HTLC that we receive must *exactly* match the information that we're provided in the onion. 

Eg: we create an invoice with final delta = 18, at height 100. Somebody goes to pay the invoice, creating a final HTLC that expires at height 118. We will not accept a HTLC for this invoice that expires at block 113, because the HTLC *must* match the delta that we gave (even though in other cases we'd accept something lower). Because the [spec](https://github.com/lightning/bolts/blob/master/04-onion-routing.md#requirements) says: 
`MUST return an error if incoming cltv_expiry < current_block_height + min_final_cltv_expiry_delta.`

> But maybe some other implementations do check outgoing_cltv_value > current_block_height which is not part of the spec. So we could just add a 3 block padding to be absolutely sure.

I think we should add the padding so that we have better payment success - if a block is mined during payment, we need to have the delta or the payment will fail?

ziggie1984 replied:
> The padding is not baked in. The HTLC that we receive must exactly match the information that we're provided in the onion.

Ohh you are right, did only think of `sendpayment` when analysing this. For the query route command we do not add the blockpadding when not using blinded paths. Maybe worth adding it in a separate PR but really minor issue ?


> I think we should add the padding so that we have better payment success - if a block is mined during payment, we need to have the delta or the payment will fail?

Yes I think your approach is the right one here. We should add a padding in blinded paths as well. One could the padding on the receiver add compared to the sender but I like the approach way more on the sender side.


So everything looks good now just the minor issue we discussed where we want the ability to add a cltv_delta in case the blinded path has only 1 hop?

carlaKC replied:
> So everything looks good now just the minor issue we discussed where we want the ability to add a cltv_delta in case the blinded path has only 1 hop?

Yeah think so! Adding that in shortly. 

Re block padding: going to stay consistent with how we have it now and *not* have block padding for blinded routes (discussed offline - it's confusing to have a difference here). Instead, going to follow this PR up with a new `block_padding` field for the `QueryRoutesRequest` so that senders can opt in.

carlaKC commented on itest/lnd_route_blinding.go:
```diff
@@ -0,0 +1,288 @@
+	// the route.
+	totalFee := bobCarolBase + blindedBaseFee
+	totalAmt := uint64(paymentAmt) + totalFee
+	totalCltv := info.BlockHeight + bobCarolDelta + blindedCltvDelta
```
We can't test receiver side logic until it's implemented in LND?

carlaKC commented:
Addressed latest nits + fixed some errant proto formatting in last push. 

One outstanding q for me - are we happy to [fix path size estimation](https://github.com/lightningnetwork/lnd/pull/7267/commits/7bb1d45754731b14b91b6b1ecc50e0e3a8b68023#r1261250580) in a follow up? Rationale being that: 
1. It's borked already (doesn't include metadata, MPP fields etc)
2. We don't come near the limit in practice
3. Touching `findPath` is scary, and deserves individual attention

ziggie1984 commented on itest/lnd_route_blinding.go:
```diff
@@ -283,6 +284,41 @@ func testQueryBlindedRoutes(ht *lntest.HarnessTest) {
+		},
+		// Fees and expiry are zero for single-hop blinded paths.
+		BaseFeeMsat:    0,
+		TotalCltvDelta: 0,
```
I agree, I was just referring to the comment, this reads like it's always the case for `single_hop blinded paths`, suggestion:

`// Fees and expiry are set to zero for this single-hop blinded paths testcase`

carlaKC replied:
Ahyea, I'll fix on next push!

carlaKC commented:
Latest push is just a [rebase](https://github.com/lightningnetwork/lnd/compare/7c7c1f4fec3a0c8a9a315bc42bfe2c7d56694946..7e948eb79ae1f2ecf3c1fc499624e77eedabd39b) on master to clear the flake!

ziggie1984 commented on cmd/lncli/cmd_payments.go:
```diff
@@ -99,6 +99,45 @@ var (
 		Usage: "(optional) expresses time preference (range -1 to 1)",
 	}
+
+	introductionNodeFlag = cli.StringFlag{
```
I think it makes sense to keep this commit maybe until we have the possibility to pay directly to bolt12 invoices ? Would also be practical when testing lnd forwarding of blinded payments which the next PR introduces I think? And when the functionality is ready drop this commit?

carlaKC commented:
tl;dr for horrified reviewers coming back to this: 
* No _major_ changes to the structure of the PR, more tests!
* Fixed an edge case where we would not include the final cltv delta for a single hop blinded route (ie, direct to the introduction node). 

Decided to add two small followups here: 
1. Add an optional `block_padding` to `QueryRoutesRequest` so that we can account for the possibility that a block is mined while we're pathfinding. Seems like it would be generally a good idea for the API, allows us to be consistent with blinded / non-blinded queries.
2. Better hop [size estimation](https://github.com/lightningnetwork/lnd/blob/bd3f570107244688583f450653e906943f69a2f4/routing/pathfind.go#L773) in `findPath`:  we need to account for encrypted data + blinding point here, since touching `findPath` is sensitive we'll quickly follow with a change just focused on this.

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@ellemouton: review reminder
@bitromortac: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@ellemouton: review reminder
@bitromortac: review reminder
@carlakc, remember to re-request review from reviewers when ready

Roasbeef commented:
Useful link for reviewers looking to catch up with the latest revisions here: https://reviewable.io/reviews/lightningnetwork/lnd/7267

carlaKC commented:
Latest push just a rebase - ready for review when it's available 🙏

carlaKC commented:
<p align="center">
<img width="585" alt="Good news everybody" src="https://github.com/lightningnetwork/lnd/assets/42311294/117317e4-5c46-4ee0-a01d-89e2b886c85b">
</p>

Interop with LDK achieved! 

Final [push](https://github.com/lightningnetwork/lnd/compare/4d028e69798d91cba3cd15543c9f8ff80f938edd..be9e8aaad2c4424af87953165e97c3ba6b9763b9) adds the new `total_amt_msat` TLV to the final payload, which is required for all blinded routes. 

Massive h/t to @bitromortac who pointed this out in the review, but I incorrectly didn't think we needed it (I have paid for this hubris with an extra hour of interop debugging).

Details of various commits for the record/ those looking to reproduce:

- LDK @ [ac5c56c](https://github.com/carlaKC/rust-lightning/commit/ac5c56c2734a866a09b6212d3ee1fb3e15d84160)
- LND @ [22b1aae1](https://github.com/lightningnetwork/lnd/pull/7267/commits/22b1aae16663f7b83807efa7064f4b9e0799eb2e)
- LDK-Sample @ [94d1705](https://github.com/carlaKC/ldk-sample/commit/94d1705e1f52bdc35cef42e605ca014b23280d5a)

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@ellemouton: review reminder
@bitromortac: review reminder
@carlakc, remember to re-request review from reviewers when ready

yyforyongyu commented on itest/lnd_route_blinding.go:
```diff
@@ -0,0 +1,290 @@
+	ctxt, cancel := context.WithTimeout(ctxb, defaultTimeout)
+	defer cancel()
+
+	info, err := alice.RPC.LN.GetInfo(ctxt, &lnrpc.GetInfoRequest{})
```
we now have a new set of methods, so can just all `alice.RPC.GetInfo` here

carlaKC replied:
<img width="452" alt="Screenshot 2023-08-24 at 9 32 44 AM" src="https://github.com/lightningnetwork/lnd/assets/42311294/7516c256-3cec-4b55-8a69-047ca069e752">

yyforyongyu commented on itest/lnd_route_blinding.go:
```diff
@@ -0,0 +1,290 @@
+// expected.
+func testQueryBlindedRoutes(ht *lntest.HarnessTest) {
+	var (
+		ctxb = context.Background()
```
hmm we don't need to init context anymore...

yyforyongyu commented on itest/lnd_route_blinding.go:
```diff
@@ -0,0 +1,290 @@
+	}
+
+	r := resp.Routes[0]
+	assert.Equal(ht, expectedRoute.TotalTimeLock, r.TotalTimeLock)
```
use `require`?

carlaKC replied:
For this test I've found it more useful to assert on the values so that I can see _everything_ that went wrong (rather than killing the test on the first bad value) because it gives a clearer indication of why it's failing. 

Happy to change, but I do think it's valuable to `assert` sometimes.

yyforyongyu replied:
oh nice I didn't know that! Yeah that's very helpful.

yyforyongyu reviewed (commented):
sorry a bit late in the game🙈 I think this is veeeery close🤖

carlaKC commented on routing/blinding.go:
```diff
@@ -67,3 +69,89 @@ func (b *BlindedPayment) Validate() error {
+	// an introduction node (this is a valid path according to the spec).
+	// Since we have the un-blinded node ID for the introduction node, we
+	// don't need to add any route hints.
+	if len(b.BlindedPath.BlindedHops) <= 1 {
```
We're allowed a blinded path with length 1 (only <1 fails validation). Will change this to `=1` so it's clearer.

carlaKC commented on routing/pathfind.go:
```diff
@@ -260,6 +274,47 @@ func newRoute(sourceVertex route.Vertex,
+			)
+		)
+
+		for i, hop := range hops {
```
Yeah, I just think it's _way_ more readable like this. 

I tried an approach that puts it in the loop above, but it gets pretty gnarley when we're zero-ing out the amount/timelock but still calculating totals etc.

yyforyongyu commented on cmd/lncli/cmd_payments.go:
```diff
@@ -1144,6 +1201,89 @@ func queryRoutes(ctx *cli.Context) error {
+
+	// If none of the blinded route params are set as we expect, then we
+	// don't need to return anything.
+	haveBlinded := ctx.IsSet(blindingPointFlag.Name) ||
```
should checking only for `blindingPointFlag` be sufficient?

yyforyongyu commented on cmd/lncli/cmd_payments.go:
```diff
@@ -1123,16 +1174,22 @@ func queryRoutes(ctx *cli.Context) error {
-		IgnoredPairs:      ignoredPairs,
+		PubKey:              dest,
+		Amt:                 amt,
+		FeeLimit:            feeLimit,
```
should we check this is not set for blinded routes?

carlaKC replied:
This logic should apply as before for blinded paths?

ziggie1984 replied:
exactly this feelimit still needs to be in place for blinded routes as well.

yyforyongyu commented on cmd/lncli/cmd_payments.go:
```diff
@@ -1054,6 +1093,12 @@ var queryRoutesCommand = cli.Command{
 		},
 		timePrefFlag,
 		cltvLimitFlag,
+		introductionNodeFlag,
```
how about creating a new cli, sth like `queryrouteswithblindedpath`?

yyforyongyu commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -258,34 +256,132 @@ func (r *RouterBackend) parseQueryRoutesRequest(in *lnrpc.QueryRoutesRequest) (
-	}
+	// Validate that the fields provided in the request are sane depending
+	// on whether it is using a blinded path or not.
+	if len(in.BlindedPaymentPaths) > 0 { //nolint:nestif
```
If we create new methods for this we can get rid of the two linters, otherwise it kinda defeats the purpose of having them.

carlaKC commented on routing/router.go:
```diff
@@ -1803,16 +1803,94 @@ type routingMsg struct {
+	return *r.Target
+}
+
+func (r *RouteRequest) hints() RouteHints {
```
These methods are introduced because we're using a single request to represent regular and blinded payments. 

Take target as an example: 
- For a regular payment we provide a pubkey.
- For a blinded payment we extract the final pubkey from the blinded route. 

The method is a way for callers to interact with the multi-purpose request without needing to worry about those details. Things like `Amount` are different, because we use use that value as-is. 

I made them private because they're not needed outside of `router`, but I agree it's kind of confusing. I also thought about interfacing this out, but it ended up much more boiler-plate-ey. 

For now, going to update this with clearer comments and unexport the fields that we use in the private methods. 
Open to suggestions on how to improve this!

carlaKC commented on routing/blinding.go:
```diff
@@ -67,3 +69,89 @@ func (b *BlindedPayment) Validate() error {
+// path.
+func (b *BlindedPayment) toRouteHints() map[route.Vertex][]*channeldb.CachedEdgePolicy { //nolint:lll
+	// If we just have a single hop in our blinded route, it just contains
+	// an introduction node (this is a valid path according to the spec).
```
Not sure I'm following - could you elaborate?

The introduction node will still need to do it's route blinding duties (when it receives a blinding point in its onion payload), we just don't need to tell pathfinding that there are any hops after it?

carlaKC commented on routing/pathfind.go:
```diff
@@ -260,6 +274,47 @@ func newRoute(sourceVertex route.Vertex,
+		for i, hop := range hops {
+			// Once we locate our introduction node, we know that
+			// every hop after this is part of the blinded route.
+			if bytes.Equal(hop.PubKeyBytes[:], introVertex[:]) {
```
Since we're ranging through hops forward, we'll always hit the `IntroductionPoint` first here (I think it's fair to trust our pathfinding to not include one node more than once).

> guess because e(i+1) = H(E(i) || ss(i)) * e(i), e(i+1)

But yeah, well formed blinded routes will be unique!

carlaKC commented on routing/pathfind.go:
```diff
@@ -260,6 +274,47 @@ func newRoute(sourceVertex route.Vertex,
+				continue
+			}
+
+			payload := blindedPath.BlindedHops[dataIndex].CipherText
```
I think it's fair to assume that we have a valid blinded path by the time we get here? 
Will add a comment to note this expectation.

carlaKC commented on cmd/lncli/cmd_payments.go:
```diff
@@ -99,6 +99,45 @@ var (
 	}
+
+	introductionNodeFlag = cli.StringFlag{
+		Name: "introduction_node",
```
How about `(blinded paths)` rather than optional? I think that could be a nice way to visually distinguish it.

carlaKC commented:
Pushed changes refactor: 
* `NewRouteRequest`: set fields that are common to regular and blinded payments in the constructor rather than having confusing private methods + added some test coverage for that. 
* `lnrpc/parseQueryRoutes`: broke up into more function to remove `nolint` (very basic change, I was being lazy).
* `lncli/queryroutes`: prefix with `blinded paths` to make params more distinguishable. 

Also added docs in `release-notes/18` with minimal boiler plate (assume that desired structure will be added in a separate PR).

yyforyongyu commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -353,22 +409,144 @@ func (r *RouterBackend) parseQueryRoutesRequest(in *lnrpc.QueryRoutesRequest) (
+		return nil, err
+	}
+
+	if len(rpcPath.BlindedHops) < 1 { //nolint:gomnd
```
think we've disabled `gomnd` after a few complains😂

carlaKC replied:
😈

yyforyongyu commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -460,7 +467,26 @@ func UnmarshallHopWithPubkey(rpcHop *lnrpc.Hop, pubkey route.Vertex) (*route.Hop
+		}
+	}
+
+	if haveBlindingPoint && len(rpcHop.EncryptedData) == 0 {
```
Q: if we only have one blinded hop, do we still need this `EncryptedData` field?

carlaKC replied:
We still need it for a single hop blinded path because it allows the recipient to verify that they originally created the blinded path (and are not being probed). 

The payload will contain a `path_id` that the introduction node/recipient will check against its invoices (kind of like a `payment_secret` in today's world).

yyforyongyu reviewed (commented):
I have one final thing re empty `BlindedHops`, otherwise it's good from my side🏖 Thanks again for the well-formatted commits, they are extremely reviewer-friendly🙏

carlaKC commented on routing/router.go:
```diff
@@ -1828,39 +1860,131 @@ type RouteRequest struct {
+		requestExpiry = finalExpiry
+	)
+
+	requestTarget, err := getTargetNode(target, blindedPayment)
```
Ah yeah, nice catch! Moving down + adding coverage for an invalid blinded path.

carlaKC commented on routing/router.go:
```diff
@@ -3034,7 +3044,7 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
 			records:     nil,
 			paymentAddr: payAddr,
-		},
+		}, nil,
```
Yeah, just starting with `QueryRoutes` as it was the smallest cut! 

I think we'll only want to start looking at the other APIs (like `SendPayment`) when LND understands invoices that contain blinded paths (which is prob a while off).

carlaKC commented:
Just the [nil blinded payment fix/test](https://github.com/lightningnetwork/lnd/compare/81632c06761cbc2cc15d537e5023f2d7cecb5812..c745d29ac2f6637a6991cb063672eebe4339a771) in last push!

yyforyongyu reviewed (approved):
LGTM🏖🌊⛵️

ziggie1984 commented on cmd/lncli/cmd_payments.go:
```diff
@@ -1144,6 +1203,82 @@ func queryRoutes(ctx *cli.Context) error {
+	}
+
+	// If any one of our blinding related flags is set, we expect the
+	// full set to be set and we'll error our accordingly.
```
Nit s/error our/error out/g

ziggie1984 reviewed (approved):
LGTM ⭐️

carlaKC commented on routing/blinding.go:
```diff
@@ -77,3 +79,94 @@ func (b *BlindedPayment) Validate() error {
+	// route as the policy for the hint from our introduction node. This
+	// will ensure that pathfinding provides sufficient fees/delay for the
+	// blinded portion to the introduction node.
+	introductionNode := b.BlindedPath.BlindedHops[1].BlindedNodePub
```
indeed!

carlaKC commented:
Thanks for the review @ziggie1984 ! [Fixed up](https://github.com/lightningnetwork/lnd/compare/c745d29ac2f6637a6991cb063672eebe4339a771..1476e416963f24f8d168655f6fc73adbbf724c4d) last few typos/ var names as suggested in latest push.

yyforyongyu commented:
Changed base to 0.18, now needs a rebase!

carlaKC commented:
Please give me a shout when this needs a rebase! 
Seems unnecessarily noisy to rebase on docs/protos so I'm going to leave as-is for now.

lightninglabs-deploy commented:
@ellemouton: review reminder
@bitromortac: review reminder
@carlakc, remember to re-request review from reviewers when ready

yyforyongyu commented:
let's rebase and merge? cc @guggero

lightninglabs-deploy commented:
@ellemouton: review reminder
@bitromortac: review reminder
@carlakc, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@ellemouton: review reminder
@bitromortac: review reminder
@carlakc, remember to re-request review from reviewers when ready

carlaKC commented:
> the error will be converted to a normal error by the introduction node if I understand correctly, which means that they will be seen as the node to blame. I think that this could lead to a decreased incentive for routers to support route blinding. 

Yeah def! I owe ya'll a quick mission control follow up for this PR (discussed somewhere in here but I've lost it in the comments). Been thinking about this more and I think that it makes sense to penalize every node _after_ the introduction node:
- Intro node has no incentive to "fake blame" everyone else because they are not permanently affected (since they're blinded), and the intro getting paid depends on the blinded route. 
- That will have the effect of discounting the blinded path (which starts to matter when we have multiple blinded paths).
